File: src/commentbox/CommentBox.js
import React, {Component} from 'react';
import Comment from './Comment';
import CommentForm from './CommentForm';


class CommentBox extends Component {

    constructor() {
        super();

        this.state = {
            showComments: false,
            commentList: []
        };

        this._getComments = this._getComments.bind(this);
        this._handleClick = this._handleClick.bind(this);
        this._addComment = this._addComment.bind(this)
    }

    _getComments() {
        return this.state.commentList.map((comment) => {
            return <Comment author={comment.author} comment={comment.comment}/>
        })
    }

    _addComment(author, comment) {
        this.setState({
         commentList: this.state.commentList.concat([{author, comment}])})
        this.props.updateNoOfComments(this.state.commentList.length)
    }

    _handleClick() {
        this.setState({
            showComments: !this.state.showComments
        })
    }

    render() {
        let commentNodes;
        let buttonText = "Show Comments";
        const comments = this._getComments();
        if (this.state.showComments) {
            buttonText = "Hide Comments";
            commentNodes = <div className="comment-list">{comments}</div>
        }

        return (
            <div className="comment-box">
                <h3>Comments</h3>
                < button onClick={this._handleClick}> {buttonText} </button>
                {commentNodes}
                <CommentForm add={this._addComment}/>
            </div>
        );
    }
}

export default CommentBox;

File: src/main/scala/samples/CommentListExample/commentList/CommentBox.scala
package samples.CommentListExample.commentList

import com.github.ahnfelt.react4s._
import samples.CommentListExample.commentList.CommentBox.UpdateNoOfComments
import samples.CommentListExample.commentList.CommentForm.AddComment

case class CommentBox() extends Component[CommentBox.Msg] {

  val showCommentsS  = State(false)
  val commentModelsS = State(List.empty[CommentModel])

  override def render(get: Get): ElementOrComponent = {
    val showComments                      = get(showCommentsS)
    val commentModels: List[CommentModel] = get(commentModelsS)

    val buttonText = if (showComments) "Hide Comments" else "Show Comments"

    val commentNodes = if (showComments) {
      E.ul(Tags(commentModels.map { comment =>
        E.li(Component(Comment, comment.author, comment.comment))
      }))
    } else {
      E.ul()
    }

    def handleCommentFormEvents(msg: CommentForm.Msg): Unit = msg match {
      case AddComment(author, comment) =>
        val commentModels = get(commentModelsS)
        commentModelsS.set(commentModels :+ CommentModel(author, comment))
        emit(UpdateNoOfComments(commentModels.length))
    }

    E.div(
      E.h3(Text("Comments")),
      E.button(Text(buttonText), A.onClick { e =>
        e.preventDefault()
        showCommentsS.set(!showComments)
      }),
      commentNodes,
      E.div(Component(CommentForm).withHandler(handleCommentFormEvents))
    )
  }

}

object CommentBox {
  sealed trait Msg
  case class UpdateNoOfComments(noOfComments: Int) extends Msg

}

case class CommentModel(author: String, comment: String)
